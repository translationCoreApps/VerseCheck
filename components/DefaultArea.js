import React from 'react'
import {Glyphicon} from 'react-bootstrap'
import style from '../css/Style'
import {selectionArray, occurrencesInString, normalizeString} from '../utils/selectionHelpers'
import MyLanguageModal from './MyLanguageModal'

class DefaultArea extends React.Component {
  constructor() {
    super();
    this.state = {
      inBox: false,
      modalVisibility: false
    }
  }

  displayText(verseText, selections) {
    // normalize whitespace for text rendering in order to display highlights with more than one space since html selections show one space
    verseText = normalizeString(verseText);
    let verseTextSpans = <span>{verseText}</span>;
    let selectMode = this.props.mode === "select";
    if (selections && selections.length > 0) {
      let _selectionArray = selectionArray(verseText, selections);
      selections.forEach(selection => {
        if (occurrencesInString(verseText,selection.text) !== selection.occurrences) {
          // validate selections and remove ones that do not apply
          this.props.actions.validateSelections(verseText);
        }
      })
      verseTextSpans = _selectionArray.map((selection, index) => {
        let style = selection.selected ? { backgroundColor: 'var(--highlight-color)' } : {};
        return (
          <span key={index} style={style}>
            {selection.text}
          </span>
        );
      })
    }
    return (
      <div style={{userSelect: 'none', opacity: selectMode ? '0.5' : 1}}>
        {verseTextSpans}
      </div>
    )
  }

  render() {
    const { projectDetailsReducer } = this.props;
    const { manifest, bookName } = projectDetailsReducer;
    const reference = this.props.contextIdReducer.contextId.reference;
    const bibles = this.props.resourcesReducer.bibles;
    const languageName = manifest.target_language ? manifest.target_language.name : null;
    const dir = manifest.target_language ? manifest.target_language.direction : null;

    return (
      <div style={{flex: 1, display: 'flex', flexDirection: 'column'}}>
        <div style={style.verseTitle}>
          <div style={{display: 'flex', flexDirection: 'column'}}>
              <span style={style.pane.title}>
                {languageName}
              </span>
              <span style={style.pane.subtitle}>
                {bookName} {reference.chapter + ':' + reference.verse}
              </span>
          </div>
          <div onClick={() => {
            this.setState({modalVisibility: true})
          }}>
            <Glyphicon glyph="fullscreen" title="Click to show expanded verses" style={{cursor: "pointer"}}/>
            {this.state.modalVisibility ? 
              <MyLanguageModal
                show={this.state.modalVisibility}
                targetLangBible={bibles.targetLanguage}
                chapter={reference.chapter}
                currentVerse={reference.verse}
                dir = {dir ? dir : "ltr"}
                onHide={
                  () => {
                    this.setState({modalVisibility: false})
                  }
                }
              />
              :
              <div />
            }
          </div>
        </div>
        <div style={this.props.projectDetailsReducer.params.direction === 'ltr' ? style.pane.contentLTR : style.pane.contentRTL}>
          {this.displayText(this.props.verseText, this.props.selectionsReducer.selections)}
        </div>
      </div>
    )
  }
}

export default DefaultArea;
