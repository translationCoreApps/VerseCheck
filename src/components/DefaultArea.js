import React from 'react';
// helpers
import {
  selectionArray,
  occurrencesInString,
  normalizeString
} from '../utils/selectionHelpers';
// components
import { Glyphicon } from 'react-bootstrap';
import InstructionsArea from './InstructionsArea';
import MyLanguageModal from './MyLanguageModal';
// styling
import style from '../css/Style';

class DefaultArea extends React.Component {
  constructor() {
    super();
    this.state = {
      inBox: false,
      modalVisibility: false
    };
  }

  displayText(verseText, selections) {
    // normalize whitespace for text rendering in order to display highlights with more than one space since html selections show one space
    verseText = normalizeString(verseText);
    let verseTextSpans = <span>{verseText}</span>;
    if (selections && selections.length > 0) {
      let _selectionArray = selectionArray(verseText, selections);
      selections.forEach(selection => {
        if (occurrencesInString(verseText,selection.text) !== selection.occurrences) {
          // validate selections and remove ones that do not apply
          this.props.actions.validateSelections(verseText);
        }
      });
      verseTextSpans = _selectionArray.map((selection, index) => {
        let style = selection.selected ? { backgroundColor: 'var(--highlight-color)' } : {};
        return (
          <span key={index} style={style}>
            {selection.text}
          </span>
        );
      });
    }
    return (
      <div style={{userSelect: 'none', color:'var(--text-color-light)'}}>
        {verseTextSpans}
      </div>
    );
  }

  render() {
    const {
      projectDetailsReducer: {
        manifest
      }
    } = this.props;
    const bookName = manifest.target_language.book.name;
    const reference = this.props.contextIdReducer.contextId.reference;
    const bibles = this.props.resourcesReducer.bibles;
    const languageName = manifest.target_language ? manifest.target_language.name : null;
    const dir = manifest.target_language ? manifest.target_language.direction : null;

    return (
      <div style={{WebkitUserSelect:'none', flex: 1, display: 'flex', flexDirection: 'column'}}>
        <div style={style.verseTitle}>
          <div style={{display: 'flex', flexDirection: 'column'}}>
              <span style={style.pane.title}>
                {languageName}
              </span>
              <span style={style.pane.subtitle}>
                {bookName} {reference.chapter + ':' + reference.verse}
              </span>
          </div>
          <div onClick={() => {this.setState({modalVisibility: true})}}>
            <Glyphicon glyph="fullscreen" title="Click to show expanded verses" style={{cursor: "pointer"}}/>
          </div>
          <MyLanguageModal
          projectDetailsReducer={this.props.projectDetailsReducer}
          show={this.state.modalVisibility}
          targetLangBible={bibles.targetLanguage}
          chapter={reference.chapter}
          currentVerse={reference.verse}
          dir = {dir ? dir : "ltr"}
          onHide={
            () => {
              this.setState({modalVisibility: false});
            }
          }/>
        </div>
        <div style={this.props.projectDetailsReducer.manifest.target_language.direction === 'ltr' ? style.pane.contentLTR : style.pane.contentRTL}>
          {this.displayText(this.props.verseText, this.props.selectionsReducer.selections)}
        </div>
      </div>
    );
  }
}

export default DefaultArea;
