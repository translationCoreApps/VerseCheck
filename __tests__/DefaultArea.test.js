/* eslint-env jest */
import React from 'react';
import DefaultArea from '../src/components/DefaultArea';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('DefaultArea component Tests', () => {
  let props;

  beforeEach(()=> {
    props = {
      actions: {
        validateSelections: jest.fn(),
      },
      contextIdReducer: {
        contextId: {
          quote: 'Θεοῦ',
          occurrence: 2,
          occurrences: 2,
          reference: {
            chapter: 1,
            verse: 1
          }
        }
      },
      resourcesReducer: {
        bibles: {
          targetLanguage: {
            targetBible: {}
          }
        }
      },
      projectDetailsReducer: {
        manifest: {
          target_language: {
            direction: 'ltr'
          },
          project: {
            name: 'Titus'
          }
        }
      },
      selectionsReducer: {
        selections: [{text:'text'}]
      },
      verseText: 'verse text'
    };
  });

  test('Check DefaultArea component', () => {
    const wrapper = shallow(<DefaultArea {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual('Titus 1:1<Glyphicon /><MyLanguageModal />verse text');
  });
});
