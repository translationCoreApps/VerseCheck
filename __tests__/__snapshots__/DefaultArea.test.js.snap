// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultArea component Tests Check DefaultArea component 1`] = `
<div
  style={
    Object {
      "WebkitUserSelect": "none",
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flex": "0 0 45px",
        "justifyContent": "space-between",
        "margin": "5px 10px 5px 15px",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
        }
      }
    >
      <span
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      />
      <span
        style={
          Object {
            "color": "var(--text-color-light)",
          }
        }
      >
        Titus
         
        1:1
      </span>
    </div>
    <div
      onClick={[Function]}
    >
      <Glyphicon
        bsClass="glyphicon"
        glyph="fullscreen"
        style={
          Object {
            "cursor": "pointer",
          }
        }
        title="Click to show expanded verses"
      />
    </div>
    <MyLanguageModal
      chapter={1}
      currentVerse={1}
      dir="ltr"
      onHide={[Function]}
      projectDetailsReducer={
        Object {
          "manifest": Object {
            "project": Object {
              "name": "Titus",
            },
            "target_language": Object {
              "direction": "ltr",
            },
          },
        }
      }
      show={false}
      targetLangBible={Object {}}
    />
  </div>
  <div
    style={
      Object {
        "direction": "ltr",
        "flex": "auto",
        "overflowY": "auto",
        "padding": "0 15px 10px",
      }
    }
  >
    <div
      style={
        Object {
          "color": "var(--text-color-light)",
          "userSelect": "none",
        }
      }
    >
      <span
        key="0"
        style={Object {}}
      >
        verse text
      </span>
      <span
        key="1"
        style={
          Object {
            "backgroundColor": "var(--highlight-color)",
          }
        }
      />
      <span
        key="2"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;
